### ISSUE TITLE
Add a LIBS configuration variable

### ISSUE THREAD

**Issue by Pawamoy** [basherpm/basher#33]
I usually have scripts inside `bin/` and libraries inside `libs/`. It would be nice if Basher also installed the file specified in LIBS inside a `lib` directory in the cellar.

**Comment by juanibiapina** [comment]
I have started this many times and failed every time because it's hard to figure out exactly what to expect of a feature like this. If we can talk and get some better clarification on how this can work, I'm up for it.

**Comment by Pawamoy** [comment]
Okay so the `LIBS` variable would specify all the files we want to install in the cellar `lib` directory. The LIBS subtrees would be respected, as `LIBS=lib/subdir/file` would be installed as `lib/subdir/file`.
By default, the whole source `lib` directory would be copied into cellar `lib` if the LIBS variable is not defined. I don't see any edge cases here.

**Comment by juanibiapina** [comment]
But what would basher do with these lib files? Just put them there and not use them? How would you use them?

**Comment by Pawamoy** [comment]
Yup I see what you mean. I'm using a global LIBPATH variable in my shell sessions, that is how I'm using the lib dir. Of course there is no standard way (or shell variable) to do this, so it would be up to the user how to use it I guess?

Personally I would simply add `~/.basher/cellar/lib` to my LIBPATH variable.

Maybe we can use another name, more generic, like bpkg `files` directive (though I'm not sure how it works), since only the user would know how to use the specified files.

**Comment by Pawamoy** [comment]
Unless Basher itself set the LIBPATH variable when doing `eval "$(basher init -)"`? Or instruct user to declare the LIBPATH variable as well?

```
export PATH="${HOME}/.basher/bin:${PATH}"
export LIBPATH="${HOME}/.basher/lib:${LIBPATH}"
eval "$(basher init -)"
```

But still there is no standard way to search and use files thanks to a colon-separated LIBPATH variable.

**Comment by Pawamoy** [comment]
For reference, it would be useful in my `shellm` project (see the [org](https://github.com/shellm-org) and [repo to be updated](https://github.com/Pawamoy/shellm)).

It provides a function called `include` to load files thanks to LIBPATH:

```bash
include() {
  local libdir array
  IFS=':' read -r -a array <<< "${LIBPATH}"
  for libdir in "${array[@]}"; do
    if [ -f "${libdir}/$1" ]; then
      # shellcheck disable=SC1090
      if ! . "${libdir}/$1"; then
        echo "shellm: include: error while including $1 (from $0)" >&2
        case "${WINDOWID}" in
          [0-9]*) [ ${SHLVL} -gt 2 ] && exit 1 || return 1 ;;
          "") [ ${SHLVL} -gt 1 ] && exit 1 || return 1 ;;
        esac
      fi
      return 0
    fi
  done
  echo "shellm: include: no such file: $1 (from $0)" >&2
  return 1
}
```

I'm currently refactoring it as a tiny core able to use other *shellm apps*, which would be installable through Basher and/or bpkg. But I need the lib files to be installed for it to work :)

**Comment by Pawamoy** [comment]
Another solution would be to execute a script called install.sh, if it exists and package.sh does not. It's what bpkg does and it's very flexible. The BASHER_PREFIX could be passed as the first arg to install.sh.

**Comment by juanibiapina** [comment]
In a way this is supported natively by shells. If the file is already in your PATH, you can just do `source file` and it will be included.

I can't see any downsides to this approach, if I understand it correctly.

**Comment by Pawamoy** [comment]
>I can't see any downsides to this approach, if I understand it correctly.

I'm not sure what approach you're talking about 👼  I said too much before 🙊  

Also, I know it's bad but I hijacked Basher to make it do what I want (see https://github.com/basherpm/basher/blob/master/libexec/basher-_unlink-completions#L11 😀 )

Here is how I wrote my `package.sh`:

```bash
BINS="bin/shellm-generate-manpages:bin/shellm-generate-wikipages"
LIBS="lib/shellman.sh"
BASH_COMPLETIONS="cmp/shellm-shellman.completion.bash"
ZSH_COMPLETIONS="cmp/shellm-shellman.completion.zsh"

DEPS=

case "$0" in
  # first run install.sh, then let run to install deps
  ${BASHER_ROOT}/libexec/basher-_deps)
    ${BASHER_PACKAGES_PATH}/shellm-org/shellm-shellman/install.sh "${PREFIX:-$BASHER_PREFIX}"
  ;;

  # first "unlink" libs, then let run
  ${BASHER_ROOT}/libexec/basher-_unlink-bins)
    IFS=: read -a libs <<< "${LIBS}"
    for lib in "${libs[@]}"; do
      rm -f "${PREFIX:-$BASHER_PREFIX}/lib/shellm/${lib#*/}"
    done
  ;;

  # let run these
  ${BASHER_ROOT}/libexec/basher-_unlink-completions) : ;;
  ${BASHER_ROOT}/libexec/basher-_unlink-man) : ;;

  # short-circuit these, already done in install.sh
  ${BASHER_ROOT}/libexec/basher-_link-bins) exit 0 ;;
  ${BASHER_ROOT}/libexec/basher-_link-completions) exit 0 ;;
  ${BASHER_ROOT}/libexec/basher-_link-man) exit 0 ;;
esac
```

**Comment by juanibiapina** [comment]
I meant the approach of sourcing files that are in path. You can just source a file which is on the bin folder. Have you tried that?

**Comment by Pawamoy** [comment]
Yes I tried, it works perfectly, but IMO there is two limitations:
- you cannot source files in subdirs (unless you include all of them in the PATH), like `source subdir/library.sh`
- and it would require:
    - at least, to mix scripts and libraries locations in the same PATH variable, which can be confusing
    - and worse, to mix scripts and libraries in the same directories (like linking lib files into Basher's bin folder)

**Comment by juanibiapina** [comment]
I don't mind linking lib files in the bin folder, as long as they are meant to be used as shell extensions (for day to day use). If a file is in the path, but is not executable, everything works fine, there is no confusion and we don't need to write any code.

Basher was designed to work only with scripts. Like I said before, I tried to include the lib feature several times, but did not find a good conceptual model to support my decisions. For now this is what I'm doing:

```sh
source "$(basher package-path juanibiapina/gg)/lib/gg.bash"
```

That's because all I need is to include this guy in the current shell for day to day use. This could also probably be solved by having a binary named `gg` and including it in the bin folder (and not making it automatically executable, so changes to basher may be required there anyway, but that's easy).

But your first point is very interesting. Indeed it doesn't work with subdirectories. Can you describe your usage in a more conceptual level? Are you trying to use basher as a runtime to load dependencies of your scripts? As you can see, currently it is only a package manager to install files, providing no runtime or other mechanisms.

**Comment by Pawamoy** [comment]
Sure!

So, the shell is my top-one productivity tool. This is what I use continuously to work. But, as you may have experienced it, I feel like everybody is just writing over and over the same things in Bash or other shells. Community efforts have been done of course (bash-it, oh-my-zsh, etc.), but most of it is only designed for interactive use (themes, completions, aliases), not really development / shell scripting (scripts and libraries). Besides, they are very monolithic (all plugins in the same repository, that you switch on and off after installation).

Basher is a solution to this, and this is why I'm willing to use it, as it allows us to write shell applications (of arbitrary sizes), and install / uninstall / update them, like any other package manager does.

The project I'm working on, shellm, is meant to be a tiny, simple library inclusion feature for the shell. On top of it, one would be able to build apps (with potential dependencies to other apps), or even projects. Think of the following analogy: `python -> django -> django-apps -> django-projects` // `shell -> shellm -> shellm-apps -> shellm-projects`. At least that's what I'm trying to achieve for myself if nobody's interested 😄 

The thing is every ecosystem needs a package manager. Python has pip, shell has basher. I don't want to reinvent the wheel, and I would be glad to promote basher in my project as the main package manager. I already wrote a [cookiecutter](https://github.com/shellm-org/cookiecutter-shellm) to bootstrap shellm apps that are installable through basher. But the distinction between scripts and libraries is very important to me, hence this issue!

As I am working a lot with Python, I take inspiration from it. I would like shellm apps to be mostly libraries, plus zero or more scripts that use them. One would be able to `basher install user/app`, then in his own app write `include app.sh`, or `include app/subset_of_functions.sh`, just like he would write `import python_app` or `from python_app import subset_of_function` in Python.

The user could also directly call packages scripts from an interactive shell with `app-script ARGS` or whatever the name is, or even `shellm app-command-name ARGS` since a main `shellm` command would look for subcommands starting with `shellm-`, just like git does. Again, it's like Python command line entry-points (or any other language scripts entry-points).

Shellm would also take care of setting the values for PATH and LIBPATH for locally and globally installed packages, always adding `~/.basher/cellar/lib` into the LIBPATH, but also `PROJECT/deps/lib` for the project you're currently working on (as well as bin directories into PATH of course).

I think I'm close to something worth with this model. Just need the package manager 😃 ! I managed to achieve this by tweaking `package.sh` as posted above, but it's hacky... I could also change the `include` function to only accept non-executable files ending with `.sh` extension, but I wanted to explore the possibility of a LIBS variable as well, as separated scripts and libraries would be cleaner I think.

**Comment by juanibiapina** [comment]
I see, you have some good points.

How about we create another tool, temporarily I'll call it `basher-include`. This tool would hook into your shell and provide you with a `include` function. It stays separate from `basher` just to maintain separation of concerns, but it knows about `basher` stuff, like for instance, the package directory structure.

I can think of some cases:

1. User never uses `include` on their interactive shell. They never need to worry about `basher-include`.
2. User uses `include` on their interactive shell: User can set up `basher-include` hook to provide that function, slowing a little bit their shell, but that's a compromise you need to make.
3. Scripts that include other scripts: They somehow use `basher-include`, which provides the `include` function. They can specify the dependencies as `DEPS=` on `basher`.

With that approach, we don't need to link the `lib` directory in basher. The `include` function could require directly from the `basher` directory structure by referring to packages by name.

If you like this plan, before we can start, we need to decide on the interface for the `include` function. How would it work?

Two possibilities from the top of my head:
```
include user/package # looks for a default include
include user/package lib/file.sh # two arguments, no magic
```

Let me know what you think.

**Comment by Pawamoy** [comment]
> 1. User never uses `include` on their interactive shell. They never need to worry about `basher-include`.

Nothing to add.

> 2. User uses `include` on their interactive shell: User can set up `basher-include` hook to provide that function, slowing a little bit their shell, but that's a compromise you need to make.

This is actually what I'm already doing. I have libraries available in my LIBPATH that I can `include` into the current shell if I want. As you can see in the following example, it doesn't cost much (not even one half of a second of loading time at shell startup):

```
       env/pyenv.sh: 0.055s
       env/goenv.sh: 0.042s
       env/rbenv.sh: 0.036s
  env/completion.sh: 0.029s
env/shellhistory.sh: 0.025s
      env/basher.sh: 0.020s
     env/aliases.sh: 0.019s
    env/sshagent.sh: 0.010s
     env/varopts.sh: 0.007s
    env/gpgagent.sh: 0.007s
    env/pydjamas.sh: 0.006s
           ext/z.sh: 0.002s
           xterm.sh: 0.002s
   core/shellman.sh: 0.001s
              cd.sh: 0.001s
        bookmark.sh: 0.001s
   env/paperless.sh: 0.001s
      env/keycut.sh: 0.001s
         env/man.sh: 0.001s
      env/prompt.sh: 0.001s
          workon.sh: 0.001s
    core/exclude.sh: 0.001s

Total load time: 0.2784 seconds
```

> 3. Scripts that include other scripts: They somehow use `basher-include`, which provides the `include` function. They can specify the dependencies as `DEPS=` on `basher`.

I always export the `include` function at shell startup (this is actually the first thing I load), so every subprocess has access to it, and scripts can use it without having to source it.

---

I don't quite like having to call `include` with user and package name, since they are related only to the location of the code on GitHub/GitLab/BitBucket. If the code moves (or is renamed), then user and package will change in basher as well, and scripts that were `include`ing it would break.

But what you said helped me think of another solution that would not require basher to install libraries, nor provide a `basher-include` command: I can simply add every `lib` directory (found in Basher installed packages) at the end of the current LIBPATH variable.

**Comment by juanibiapina** [comment]
Good that you found a simple solution that works for you!

The `user/package` pattern is used also by `cargo` and Docker Hub, among others. In case of `basher`, it would break if the package is renamed, but I see that as an advantage, because it allows two packages to have the same name, but different scopes (user).

I might still go ahead with my include plan anyway. The problem you made me see now is that if the user loads it in his shell, then individual packages wouldn't need to load it themselves. If that happens then the `include` function would be loaded twice, which is useless. I might just add include as part of `basher init -` and check what the performance difference is in practice. Then it would be available to all scripts, like you said.

Also, does shellm provide more functions besides `include` ? If I do it in `basher`, you wouldn't need to worry about coding `include` or LIBPATH, but I don't want to hijack your project.

**Comment by Pawamoy** [comment]
Currently yes, shellm provides more stuff, but I am refactoring it so it would only provide the `include` mechanism (among other closely-related stuff). Please don't hesitate to go forward with `basher-include` (or the name you would like), I would be glad to see how you did it 😄 

Also, you talked about the function being loaded twice: I am solving this issue in libraries by mimic'ing  the C libraries, i.e. with `if not_defined; then define; ...; fi`. This way each library file is ensured to be loaded only once (and we avoid infinite recursion / loop on inclusion).

**Comment by juanibiapina** [comment]
I meant "not load twice" in a sense that scripts don't need to require `basher` or `shellm` for them to work, they can assume `include` is a defined function.

I'll try a spike on this soon to see how it goes.

**Comment by Pawamoy** [comment]
Hi again!

So after a bit thinking, I think you are right about using the namespaces for packages when including them, as in `include namespace/package`.

Also, the `include user/package lib/file.sh` syntax is a bit verbose but I think it's a good compromise to keep a great flexibility. But maybe instead of using a space, we could use a slash, like `include user/package/lib/file.sh`. Indeed `lib/file.sh` is contained in `user/package` after all!

I also like the idea of searching for a default file to include when doing `include namespace/package`. Maybe one of the following `main.sh`, `lib.sh`, `all.sh` or `PACKAGE_NAME.sh`?

At first I didn't like having the inclusion mechanism as part of the package manager, but I guess both are simply too tied to each other, just like `require` in node.js knows how to load packages installed by `npm`.

To conclude I would be glad to work with you on this. I don't care if it hijacks a bit what I wanted to do with shellm since it's only the starting point. If I have access to a functional `basher-include` or `include` function, I will still be able to wrap it in shellm to enforce a coding standard/style to be used in shellm apps.

**Comment by juanibiapina** [comment]
If I get some free time this week, I'll check what's required to get this working.

**Comment by redrock9** [comment]
Hi guys, 

I saw this thread and thought that I might add something. I think that your ideas for inclusion of libraries are awesome and I'd gladly contribute to their realisation. I do, however, have something to add to this: I think that as soon as libraries come into play for a language, standards need to be created as to what a library file should look like. Nodejs modules for example have the module.exports directive, and Java has its packages/classes system to neatly (though verbosely) organize imports. A spec doesn't need to be very verbose or have a huge amount of rules, but it should enforce consistency, so that the library can be sourced/used consistently.

Because I like to source scripts with reusable functions, I like to create my own little "modules". I have been thinking of a way to make it easy to tell module functions apart from locally defined functions, and to avoid overriding already defined functions. What I came up with is the following spec:
- Module functions should be prefixed with the name of the module in PascalCase, followed by an underscore.
- Module function names should be defined in camelCase.
- Modules can not execute any code or define any variables outside of function bodies.

I have written a small module that can do commonly performed array operations using this spec, you can take a look at it here to get an idea of what I mean: https://github.com/redrock9/bash-array-utils .
It's a simple spec, but because the definition of the functions is consistent, it is a small efford to solve scoping issues when sourcing two modules with the same name, because you could just replace the module prefix. My Array module could for example be sourced as ArrayUtils like this: `source <(sed 's/Array_/ArrayUtils_/g;' Array.bash)`.  That could be wrapped in the include function by adding an optional argument to the include function, so that something like `include author/ModuleName as DesiredName` would be possible.

To ensure modules adhere to the spec, a small parser could be written that is run over the module at install time (not at time of including/importing as that would unnecessarily slow down script execution) and tells the user when a module, for instance, executes code outside of a function block or what other guidelines have not been adhered to.

What do you think of this idea?

**Comment by juanibiapina** [comment]
I just want to have the simplest thing the works, it's easy to maintain and causes less problems. I also don't want to enforce naming conventions. Let people do whatever they want with their names. This doesn't solve the function name clashes, but that's a problem with the language itself that I don't want to solve. (it could be solved with a prefix convention like yours, but I won't enforce it).

The trick for sourcing modules while renaming them seems clever, but might cause so many problems that I don't want to do it.

I've implemented the first prototype. It works, makes no assumptions and no tricks. I hope it's something that we can build upon.

**Comment by Pawamoy** [comment]
@redrock9 I like the idea of being able to dynamically change module names when importing modules, with something like `include ... as ...`, but as @juanibiapina said, I think there would be so many cases to handle that it would be a hell to support and maintain. But still, I like the idea.

However, I would change things a bit:

- not a fan of PascalCase, I would personally stick to snake_case and, more generally, google shell style guide: https://google.github.io/styleguide/shell.xml
- the underscore to separate module name from function name would then become ambiguous, so I would rather use colons: `module_name::function_name` (this is valid in bash).
- if we parse things correctly, we don't even need to enforce a particular style. The only thing we need is a standardized string (such as `_` or `::`) to separate module name from function name.

Maybe we can continue this discussion elsewhere since this will not be implemented here :)

@juanibiapina I saw your changes, it seems good 👍 !

**Comment by juanibiapina** [comment]
This is implemented in #34 

The LIBS configuration variable is still not used, but could still be implemented in the future without breaking compatibility.
